### PR TITLE
Fix of ReverseInputChannels for NHWC layout.

### PR DESCRIPTION
**Pull request by popovaan**
Root cause analysis: 
In ReverseInputChannels index of dimension to reverse is constant equal to 1 as it was expected that layout of Parameter is always NCHW. But in API 2.0 layout can be NHWC which breaks the transformation.

Solution: 
Check if Parameter node has old_api_map and map channel index to new api.

Ticket: 70164

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests - N/A
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list - N/A
* [x]  Guide on how to convert the **public** model - N/A
* [x]  User guide update - N/A

### CODE REVIEW

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 1]
```diff
@@ -50,10 +50,24 @@ class InsertReverseChannels(BackReplacementPattern):
     """
     enabled = False
 
+    @staticmethod
+    def get_index_for_api_2_0(node: Node, idx: int):
```
Comment (sadolini) [inline model-optimizer/extensions/back/ReverseInputChannels.py:54]: imho. I don't like this function name, hard to understand from name what function do. Possibly something like get_fw_index would be better.

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 2]
```diff
@@ -50,10 +50,24 @@ class InsertReverseChannels(BackReplacementPattern):
...
+        if node.has_valid('rt_info'):
+            rt_info = node.rt_info
+            if rt_info.contains('old_api_map'):
+                old_api_map_version = rt_info.get_attribute_version('old_api_map')
+                old_api_map = rt_info.info['old_api_map', old_api_map_version]
+                if 'inverse_order' in old_api_map.info:
+                    order = old_api_map.info['inverse_order']
+                    assert len(order) > idx >= 0, 'Channel index is incompatible with old_api_map: {}.'.format(idx)
+                    return list(order).index(idx)
+        return idx
```
Comment (sadolini) [inline model-optimizer/extensions/back/ReverseInputChannels.py:64]: Is it ok if node doesn't have rt_info/old_api_map? Should we output warning in this case?

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 3]
```diff
@@ -50,10 +50,24 @@ class InsertReverseChannels(BackReplacementPattern):
...
+    @staticmethod
+    def get_index_for_api_2_0(node: Node, idx: int):
+        if node.has_valid('rt_info'):
+            rt_info = node.rt_info
+            if rt_info.contains('old_api_map'):
+                old_api_map_version = rt_info.get_attribute_version('old_api_map')
+                old_api_map = rt_info.info['old_api_map', old_api_map_version]
+                if 'inverse_order' in old_api_map.info:
+                    order = old_api_map.info['inverse_order']
+                    assert len(order) > idx >= 0, 'Channel index is incompatible with old_api_map: {}.'.format(idx)
```
Comment (sadolini) [inline model-optimizer/extensions/back/ReverseInputChannels.py:62]: Could you add node name/id in assert message to detect root cause faster?

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 4]
```diff
@@ -66,10 +80,25 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                    if order is not None and order[1] == reverse_index:
+                        # At this point layout corresponds to old API (NCHW), so the axis to reverse = 1
+                        reverse_channels = ReverseChannels(graph, {'name': name + '/reverse_input_channels',
+                                                                   'axis': 1}).create_node()
+                        out_node.out_port(0).get_connection().insert_node(reverse_channels)
+                        continue
+
+            reverse_channels = ReverseChannels(graph, {'name': name + '/reverse_input_channels',
+                                                       'axis': reverse_index}).create_node()
+            parameter.out_port(0).get_connection().insert_node(reverse_channels, attributes_save_mode='source')
```
Comment (sadolini) [inline model-optimizer/extensions/back/ReverseInputChannels.py:101]: Should we still insert ReverseChannels if parameter.out_port(0) disconnected? Possibly we can check just at the begin and ignore such parameter.

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 2 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py:64]: Yes, it's okay if node doesn't have old_api_map. This means that in new API layout and type is the same as in old API. Warning is not needed.

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 1 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py]: Done.

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 4 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py:106]: I agree. Added check and continue.

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 3 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py]: Done.

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 5]
```diff
@@ -50,10 +50,26 @@ class InsertReverseChannels(BackReplacementPattern):
     """
     enabled = False
 
+    @staticmethod
+    def get_fw_index(node: Node, idx: int):
+        if node.has_valid('rt_info'):
```
Comment (sadolini) [inline model-optimizer/extensions/back/ReverseInputChannels.py:55]: I think function with less levels of if looks better like 
if not node.has_valid('rt_info'):
    return idx
etc

**Review by sadolini — APPROVED** [review]

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 5 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py]: Done.

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 6]
```diff
@@ -295,19 +333,32 @@ class ReverseChannelsPropagationUp(BackReplacementPattern):
...
+        order = node.in_port(1).data.get_value()
+        reverse_axis = reverse_channels.axis
+        if order is None:
+            return False
+        new_axis = order[reverse_axis]
+        reverse_channels.axis = new_axis
+        return ReverseChannelsPropagationUp.lift_up_through(node, reverse_channels)
+
+    @staticmethod
+    def lift_up_through(node: Node, reverse_channels: Node):
```
Comment (jane-intel) [inline model-optimizer/extensions/back/ReverseInputChannels.py:353]: nit: 
```suggestion
    def lift_up_through_zero_port_only(node: Node, reverse_channels: Node):
```

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 7]
```diff
@@ -96,8 +121,21 @@ class ReverseChannelsPropagationDown(BackReplacementPattern):
         'ShapeOf': lambda node, rc: ReverseChannelsPropagationDown.pass_rc_through_shape(node, rc),
 
         'Pad': lambda node, rc: ReverseChannelsPropagationDown.pass_rc_through(node, rc),
```
Comment (jane-intel) [inline model-optimizer/extensions/back/ReverseInputChannels.py:123]: nit: 
```suggestion
        'Pad': lambda node, rc: ReverseChannelsPropagationDown.pass_rc_through_zero_port_only(node, rc),
```

File: model-optimizer/extensions/back/ReverseInputChannels.py [hunk 8]
```diff
@@ -50,10 +50,31 @@ class InsertReverseChannels(BackReplacementPattern):
...
+            return idx
+
+        old_api_map_version = rt_info.get_attribute_version('old_api_map')
+        old_api_map = rt_info.info['old_api_map', old_api_map_version]
+        if 'inverse_order' not in old_api_map.info:
+            return idx
+
+        order = old_api_map.info['inverse_order']
+        node_name = node.soft_get('name', node.id)
+        assert len(order) > idx >= 0, \
```
Comment (jane-intel) [inline model-optimizer/extensions/back/ReverseInputChannels.py:69]: Can we please make it work for negative axis too?

There is one particular transformation which inserts ReverseInputChannels operator outside of this file -- [here](https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/front/tf/UnpackPackReverseInputChannels.py#L67)

And it is indeed possible to have negative axis from there due to 
TF docs [here](https://www.tensorflow.org/api_docs/python/tf/unstack) and [here](https://www.tensorflow.org/api_docs/python/tf/stack) 

Could you please add such a unit-test?

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 6 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py]: Renamed.

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 7 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py]: Renamed.

File: model-optimizer/extensions/back/ReverseInputChannels.py (hunk 8 above)
Comment (popovaan) [reply model-optimizer/extensions/back/ReverseInputChannels.py:76]: Done.

**Review by jane-intel — APPROVED** [review]

**Pull request merged**